### ISSUE TITLE
Session Lock Expired with AutoLockRenew when message takes 4-5mn of processing

### ISSUE THREAD

**Issue by K3UL** [Azure/azure-sdk-for-python#14699]
- **azure-servicebus 0.50.3**
- **Debian 10**
- **Python 3.8**

**Describe the bug**
I am using the AutoLockRenew and get a `Session lock expired` when the message takes a around 5 mn to process. The exception happens on calling `complete()` on the message.

I feel like this is more or less related to #11611, #12307, and #12937, but to be honest I fail to understand the whole of it and how to apply it to my case. I don't know if it might be linked or not also to another issue of mine, #9402.
I think maybe I am failing to understand how all timeout/duration properties are supposed to work together : the lock duration of the subscription, the `receiver` timeout, the `AutoLockRenewtimeout`, and the `fetch_next` timeout.

First, on the Service Bus side, in the portal, my subscription has sessions enabled, with a TTL of 30mn and a lock duration of 5mn.

Then, my code to handle messages is the following : 
```python
sub_client = self._bus_client.get_subscription(_topic, _sub)
receiver = sub_client.get_receiver(mode=ReceiveSettleMode.PeekLock,
                                   session=(session_id or NEXT_AVAILABLE), idle_timeout=1)

locker = AutoLockRenew(max_workers=1)
try:
    sub_client: SubscriptionClient
    receiver: Receiver
    messages: List[Message] = receiver.fetch_next(max_batch_size=1, timeout=1)
    if messages:
        message = messages[0]
        # Beware, for session-enabled stuff, the Autolock has to be on the receiver, not the message
        locker.register(receiver, timeout=24*3600)
        self._handle_message(message, handler)
        return message
except NoActiveSession:
    return
finally:
    locker.shutdown(wait=False)
    receiver.close()
```

*Note :* `self._handle_message` is basically a `try: ... msg.complete() except: msg.abandon()` with some logging and verifications on message properties.

When this executes for short processing, it is fine, but I have a processing that can be a bit long and when it takes 6 minutes, I get the following : 
```python
Session lock expired
Traceback (most recent call last):
  File "/app/***/common/services/bus/***.py", line 226, in _handle_message
    msg.complete()
  File "/usr/local/lib/python3.8/site-packages/azure/servicebus/common/message.py", line 329, in complete
    self._is_live('complete')
  File "/usr/local/lib/python3.8/site-packages/azure/servicebus/common/message.py", line 105, in _is_live
    raise SessionLockExpired(inner_exception=self._receiver.auto_renew_error)
azure.servicebus.common.errors.SessionLockExpired: Session lock expired
```

Thank you !

**Comment by KieranBrantnerMagee** [comment]
Hey @K3UL 
thank _you_ for surfacing this issue to us.  The funny thing is, although I'm not sure the ones you've linked were the relevant ones, there definitely were prior echoes of this.  (That's not on you btw, it's a subtle symptom, and appreciated that you'd even try to find precedent++)

A few questions.
1. Does this fail every time?
2. If you change lock duration to 3 minutes, does this improve the situation? (I realize this is a bit of a weird ask, for context one of the prior users had observed clock skew issues along the 5 minute boundary)

We'll likely be digging into this after our release goes out this week, so thanks as always for your patience and help thus far.

**Comment by K3UL** [comment]
Hi @KieranBrantnerMagee,
Thank you for the information and the suggestions. I do see it failing, if not all the time, very regularly. It is a particularly big annoyance in some of our processes where there is an e-mail sent because it gets sent twice, as the message gets processed twice (because the second time we have cache that makes it faster, otherwise it would be processed 3 times and go to deadletter I guess).

I will check with putting the lock duration as you suggested, although it does seem a bit counter-intuitive, what will happen to my messages after 6mn ? 😅 

FYI for now we implemented, a sort of local cache of processed messages to be able to spot these cases, but it will not work if we scale something, unless we use some distributed cache, which brings a lot more complexity.

So this is a big issue for us, I will be watching, and very curious about your findings !

**Comment by axisc** [comment]
@K3UL, did setting the lock duration to 3 mins fix the symptom?

**Comment by swathipil** [comment]
Hi @K3UL,

After some investigation, we found that this was a bug in v0.50.3 and has been fixed in v7.0 and later versions.

A little more detail on why we believe this was failing:

- The service imposes a 4 minute time inactive limit on amqp link before closing the connection.
- 	When the message processes for 6 minutes, as you mentioned, the connection would get closed because there is no activity happening on the amqp receiver link.
- 	Renew session lock request was performed on the amqp management link different than then amqp receiver link, so the service considered no activity had happened during 6 minutes processing on the receiver link and chose to close the connection.
- 	In v7.0 and later, [a property was added to the mgmt request messages](https://github.com/Azure/azure-sdk-for-python/blob/4bb2951b0a34471737c7bfd2ff87256408f0d6ed/sdk/servicebus/azure-servicebus/azure/servicebus/_base_handler.py#L424-L430) to tell the service to associate the  mgmt link with the receiver link, so that connection are aware that there're activities happening.

We strongly recommend you to migrate to v7 as v0.50.3 has already been deprecated. Here is the [migration guide for moving from v0.50.3 to v7.0](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/servicebus/azure-servicebus/migration_guide.md).

Thanks for your patience!

**Comment by iferencik** [comment]
Hello,

I am using v7  async and I it looks like the bug is till there. Here is my code:

```python

import os.path

from dotenv import dotenv_values
from azure.servicebus.aio import ServiceBusClient, AutoLockRenewer

import random
import asyncio
import logging
cfg = dotenv_values('../../.env')
CONNECTION_STR = cfg['SERVICE_BUS_CONNECTION_STRING_DEV']
QUEUE_NAME='data-upload-dev'
MAX_SLEEP_SECS = 300
LOCK_SECS = 10
async def random_sleep():
    sleep_secs = random.randrange(0,MAX_SLEEP_SECS,1)
    logger.info(f'Going to sleep for {sleep_secs}')
    await asyncio.sleep(sleep_secs)


async def consume():
    #auto_lock_renewer = AutoLockRenewer(max_lock_renewal_duration=10)
    async with ServiceBusClient.from_connection_string(conn_str=CONNECTION_STR, logging_enable=True) as servicebus_client:
        async with servicebus_client.get_queue_receiver(queue_name=QUEUE_NAME) as receiver: #get one message without caching
            async with receiver:

                while True:
                    received_msgs = await receiver.receive_messages(max_message_count=1, max_wait_time=5)
                    if not received_msgs:
                        logger.info(f'No more messages to process')
                        break
                    for msg in received_msgs:
                        logger.info(f'Consumer got message {str(msg)}')
                        logger.info(f'Going to lock message {str(msg)} and auto-renew every {LOCK_SECS} secs')
                        async with AutoLockRenewer() as auto_lock_renewer:
                            auto_lock_renewer.register(receiver, msg, max_lock_renewal_duration=LOCK_SECS)
                            
                            #logger.info(f'lock on message will expire at {msg.expires_at_utc} now in utc is {datetime.datetime.utcnow()}')
                            #simulate ingest using random sleep
                            await random_sleep()
                            logger.info(f'Going to complete {str(msg)} is lock expired {msg._lock_expired}')
                            await receiver.complete_message(msg)
                            logger.info(f'{str(msg)} completed')
                            



logging.basicConfig()
logger  = logging.getLogger(os.path.split(__file__)[1])
logger.setLevel(logging.INFO)
asyncio.run(consume())
```
This systematically fails 

```bash
INFO:sbus_consumer.py:Consumer got message Message no: 8
INFO:sbus_consumer.py:Going to lock message Message no: 8 and auto-renew every 10 secs
INFO:sbus_consumer.py:Going to sleep for 75
INFO:sbus_consumer.py:Going to complete Message no: 8 is lock expired True
Traceback (most recent call last):
  File "/work/py/geohub-data-pipeline/ingest/test/sbus_consumer.py", line 51, in <module>
    asyncio.run(consume())
  File "/usr/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/work/py/geohub-data-pipeline/ingest/test/sbus_consumer.py", line 42, in consume
    await receiver.complete_message(msg)
  File "/home/janf/.local/share/virtualenvs/geohub-data-pipeline-pYPPTnlH/lib/python3.8/site-packages/azure/servicebus/aio/_servicebus_receiver_async.py", line 839, in complete_message
    await self._settle_message_with_retry(message, MESSAGE_COMPLETE)
  File "/home/janf/.local/share/virtualenvs/geohub-data-pipeline-pYPPTnlH/lib/python3.8/site-packages/azure/servicebus/aio/_servicebus_receiver_async.py", line 479, in _settle_message_with_retry
    raise ServiceBusError(
azure.servicebus.exceptions.ServiceBusError: The lock on the message lock has expired.
```
